refactor(dev): use actor design pattern and allow to recover from intial build error

hyf0 opened the pull request:
Fixes the case of recovering from intial build error at https://github.com/rolldown/rolldown/issues/6648.

- I removed `Mutex<BuildState>` to prevent potential dead lock problem.
- No shared state observation, every actor maintains its own state within itself. State transition are done by messaging.
- All operations are done by passing messages(will revise this decision for performance intensive scenario.
- General error handling is still in investigation.

hyf0 commented:
* **#6936** <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6936?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/rolldown/rolldown/6936?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6936).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/rolldown_binding/src/binding_bundler_impl.rs:
This file doesn't seem to be referenced. Is this going to be used later?

hyf0 replied:
Should be a git error

hyf0 replied:
Deleted now.

sapphi-red commented on crates/rolldown/src/dev/types/task_input.rs:
```diff
@@ -4,6 +4,8 @@ use rolldown_utils::indexmap::FxIndexSet;
+  /// Retry initial build after failure
+  InitialBuild,
   /// Initial full rebuild from scratch
   FullRebuild,
```
It seems `FullRebuild` is only created at this line: https://github.com/rolldown/rolldown/blob/eee2b98e0f6ff8abd09d22bf258434686ab2e4c3/crates/rolldown/src/dev/bundle_coordinator.rs#L65
Do we need to differentiate `InitialBuild` and `FullRebuild`? It feels like it's the same concept.

hyf0 replied:
Yes. Think again having `FullRebuild` is enough.

sapphi-red commented on crates/rolldown/src/dev/types/bundling_status.rs:
```diff
@@ -0,0 +1,14 @@
+#[derive(Debug, Clone)]
+pub struct BundlingStatus {
+  /// Whether a build is currently running
+  pub is_build_running: bool,
```
It seems this is not used anywhere.

hyf0 replied:
Removed

sapphi-red commented on crates/rolldown/src/dev/bundle_coordinator.rs:
```diff
@@ -0,0 +1,414 @@
+  }
+
+  #[expect(clippy::unused_async)]
+  async fn has_latest_build_output(&self) -> bool {
```
Why do we need `#[expect(clippy::unused_async)]` here?

hyf0 replied:
Solved

hyf0 commented on packages/test-dev-server/tests/fixtures/.edit-reload/package.json:
@sapphi-red I skipped this test and will fix it in the latter PR.

hyf0 commented:
### Merge activity

* **Nov 11, 12:47 PM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Nov 11, 12:47 PM UTC**: `hyf0` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=rolldown&repo=rolldown).
* **Nov 11, 1:17 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=rolldown&repo=rolldown).

hyf0 committed:
refactor(dev): use actor design pattern and allow to recover from intial build error (#6936)

Fixes the case of recovering from intial build error at https://github.com/rolldown/rolldown/issues/6648.

- I removed `Mutex<BuildState>` to prevent potential dead lock problem.
- No shared state observation, every actor maintains its own state within itself. State transition are done by messaging.
- All operations are done by passing messages(will revise this decision for performance intensive scenario.
- General error handling is still in investigation.
